File: app/src/main/java/com/example/messenger/ChatViewModel.java
package com.example.messenger;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;
import java.util.List;

public class ChatViewModel extends ViewModel {
    private final MutableLiveData<String> error = new MutableLiveData<>();
    private final MutableLiveData<Boolean> messageSent = new MutableLiveData<>();
    private final MutableLiveData<List<Message>> messages = new MutableLiveData<>();
    private final MutableLiveData<User> userOther = new MutableLiveData<>();

    private final FirebaseDatabase database = FirebaseDatabase.getInstance();
    private final DatabaseReference referenceUsers = database.getReference("Users");
    private DatabaseReference referenceMessages = database.getReference("Messages");

    private String currentUserId;
    private String otherUserId;

    public ChatViewModel(String currentUserId, String otherUserId) {
        this.currentUserId = currentUserId;
        this.otherUserId = otherUserId;
        referenceUsers.child(otherUserId).addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                User user = snapshot.getValue(User.class);
                userOther.setValue(user);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
        referenceMessages.child(currentUserId).child(otherUserId)
                .addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot snapshot) {
                List<Message> messageList = new ArrayList<>();
                for(DataSnapshot dataSnapshot : snapshot.getChildren()) {
                    Message message = dataSnapshot.getValue(Message.class);
                    messageList.add(message);
                }
                messages.setValue(messageList);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });
    }

    public LiveData<String> getError() {
        return error;
    }

    public LiveData<Boolean> getMessageSent() {
        return messageSent;
    }

    public LiveData<List<Message>> getMessages() {
        return messages;
    }

    public LiveData<User> getUserOther() {
        return userOther;
    }

    public void sendMessage(Message message) {
        referenceMessages
                .child(message.getSenderId())
                .child(message.getReceiverId())
                .push()
                .setValue(message)
                .addOnSuccessListener(unused ->
                        referenceMessages
                        .child(message.getReceiverId())
                        .child(message.getSenderId())
                        .push()
                        .setValue(message)
                        .addOnSuccessListener(unusedVoid -> messageSent.setValue(true))
                                .addOnFailureListener(e -> error.setValue(e.getMessage())))
                .addOnFailureListener(e -> error.setValue(e.getMessage()));
    }
}

File: app/src/main/java/com/example/messenger/ChatViewModel.kt
package com.example.messenger

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.messenger.pojo.Message
import com.example.messenger.pojo.User
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class ChatViewModel(val currentUserId: String, val otherUserId: String) : ViewModel() {
    private val error = MutableLiveData<String>()
    private val messageSent = MutableLiveData<Boolean>()
    private val messages = MutableLiveData<List<Message>>()
    private val userOther = MutableLiveData<User>()

    private val database = FirebaseDatabase.getInstance()
    private val referenceUsers = database.getReference("Users")
    private val referenceMessages = database.getReference("Messages")

    init {
        referenceUsers.child(otherUserId).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                userOther.value = snapshot.getValue(User::class.java)
            }

            override fun onCancelled(error: DatabaseError) {
                TODO("Not yet implemented")
            }
        })

        referenceMessages.child(otherUserId).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                val messageList = mutableListOf<Message>()
                for (item in snapshot.children) {
                    val message = item.getValue(Message::class.java) ?: return
                    messageList.add(message)
                }
                messages.value = messageList
            }

            override fun onCancelled(error: DatabaseError) {
                TODO("Not yet implemented")
            }
        })
    }

    fun onlineUserStatus(isOnline: Boolean) {
        referenceUsers.child(currentUserId).child("online").setValue(isOnline)
    }

    val getError: LiveData<String> get() = error
    val getMessageSent: LiveData<Boolean> get() = messageSent
    val getMessages: LiveData<List<Message>> get() = messages
    val getUserOther: LiveData<User> get() = userOther

    fun sendMessage(message: Message) {
        with(message) {
            referenceMessages
                .child(senderId)
                .child(receiverId)
                .push()
                .setValue(this)
                .addOnSuccessListener {
                    referenceMessages
                        .child(receiverId)
                        .child(senderId)
                        .push()
                        .setValue(this)
                }
        }
    }
}